pkg/provisioning/import.go

```diff
@@ -0,0 +1,712 @@
+		_, err = a.connectorService.AddProcessor(ctx, a.cfg.ID, proc.ID)
+		if err != nil {
+			return cerrors.Errorf("failed to add processor %v: %w", proc.ID, err)
+		}
+	}
+
+	return nil
+}
+func (a createConnectorAction) Rollback(ctx context.Context) error {
+	err := a.connectorService.Delete(ctx, a.cfg.ID, a.pluginService)
+	if !cerrors.Is(err, connector.ErrInstanceNotFound) {
+		// ignore instance not found errors, this means the action failed to
+		// create the connector in the first place
+		return err
+	}
+	return nil
+}
+func (createConnectorAction) connectorType(t string) connector.Type {
+	switch t {
+	case config.TypeSource:
+		return connector.TypeSource
+	case config.TypeDestination:
+		return connector.TypeDestination
+	}
+	return 0 // this is an invalid connector type, it will fail if we try to create it
+}
+
+type createProcessorAction struct {
+	cfg    config.Processor
+	parent processor.Parent
+
+	processorService ProcessorService
+}
+
+func (a createProcessorAction) String() string {
+	return fmt.Sprintf("create processor with ID %v", a.cfg.ID)
+}
+func (a createProcessorAction) Do(ctx context.Context) error {
+	_, err := a.processorService.Create(
+		ctx,
+		a.cfg.ID,
+		a.cfg.Type,
+		a.parent,
+		processor.Config{
+			Settings: a.cfg.Settings,
+			Workers:  a.cfg.Workers,
+		},
+		processor.ProvisionTypeConfig,
+	)
+	if err != nil {
+		return cerrors.Errorf("failed to create processor: %w", err)
+	}
+	return nil
+}
+func (a createProcessorAction) Rollback(ctx context.Context) error {
+	err := a.processorService.Delete(ctx, a.cfg.ID)
+	if !cerrors.Is(err, processor.ErrInstanceNotFound) {
+		// ignore instance not found errors, this means the action failed to
+		// create the processor in the first place
+		return err
+	}
+	return nil
+}
+
+// --------------------
+// -- UPDATE ACTIONS --
+// --------------------
+
+type updatePipelineAction struct {
+	oldConfig config.Pipeline
+	newConfig config.Pipeline
+
+	pipelineService PipelineService
+}
+
+func (a updatePipelineAction) String() string {
+	return fmt.Sprintf("update pipeline with ID %v", a.oldConfig.ID)
+}
+func (a updatePipelineAction) Do(ctx context.Context) error {
+	return a.update(ctx, a.newConfig)
+}
+func (a updatePipelineAction) Rollback(ctx context.Context) error {
+	return a.update(ctx, a.oldConfig)
+}
+func (a updatePipelineAction) update(ctx context.Context, cfg config.Pipeline) error {
+	p, err := a.pipelineService.Update(ctx, cfg.ID, pipeline.Config{
+		Name:        cfg.Name,
+		Description: cfg.Description,
+	})
+	if err != nil {
+		return cerrors.Errorf("failed to update pipeline: %w", err)
+	}
+	_, err = a.pipelineService.UpdateDLQ(ctx, cfg.ID, pipeline.DLQ{
+		Plugin:              cfg.DLQ.Plugin,
+		Settings:            cfg.DLQ.Settings,
+		WindowSize:          *cfg.DLQ.WindowSize,
+		WindowNackThreshold: *cfg.DLQ.WindowNackThreshold,
+	})
+	if err != nil {
+		return cerrors.Errorf("failed to update pipeline DLQ: %w", err)
+	}
+
+	// update connector IDs
```

Comment:
Is this literally because of the connector IDs themselves, or is it actually about connectors? Also, isn't this supposed to be covered through actions?

Reply:
It's just the IDs, it has nothing to do with the connectors or processors behind these IDs. Each entity (pipeline, connector, processor) is treated completely in isolation in its own action.